File: core/src/main/java/net/year4000/mapnodes/regions/CubeRegion.java
/*
 * Copyright 2016 Year4000. All Rights Reserved.
 */

package net.year4000.mapnodes.regions;

import com.flowpowered.math.vector.Vector3i;
import net.year4000.utilities.Conditions;

import java.util.Optional;
import java.util.Set;

/** Represents a cube region that wraps a cuboid region */
public final class CubeRegion extends AbstractRegion {
    private CuboidRegion region;

    /** Create a cube region with a center value and height radius and width radius */
    public CubeRegion(Vector3i center, int height, int width) {
        Conditions.nonNull(center, "center");
        Conditions.isLarger(height, 0);
        Conditions.isLarger(width, 0);
        Vector3i posOne = new Vector3i(center.getX() - width, center.getY() - height, center.getZ() - width);
        Vector3i posTwo = new Vector3i(center.getX() + width, center.getY() + height, center.getZ() + width);
        this.region = new CuboidRegion(posOne, posTwo);
    }

    /** Create a cube region with a center value and radius */
    public CubeRegion(Vector3i center, int radius) {
        this(center, radius, radius);
    }

    @Override
    public Optional<Set<Vector3i>> getPoints() {
        return region.getPoints();
    }

    @Override
    public boolean contains(Vector3i vector3i) {
        return region.contains(vector3i);
    }
}

File: core/src/main/js/mapnodes/regions/cube_region.js
/*
 * Copyright 2020 Year4000. All Rights Reserved.
 */
import _ from 'lodash'
import { Vector3 } from 'three'

import AbstractRegion from './abstract_region.js'
import CuboidRegion from './cuboid_region.js'
import { not_null } from '../conditions.js'


/** Represents a cube region that wraps a cuboid region */
export default class CubeRegion extends AbstractRegion {
  /**
   * Center is a vector3, when width is not there height becomes radius
   *
   * @param {Vector3} center
   * @param {number} height
   * @param {number} width
   */
  constructor(center, height, width) {
    super()
    not_null(center, 'center')
    not_null(height, 'height')
    width = width || height
    const point_one = new Vector3(center.x - width, center.y - height, center.z - width)
    const point_two = new Vector3(center.x + width, center.y + height, center.z + width)
    this._cuboid = new CuboidRegion(point_one, point_two)
  }

  /** Get all the immutable set of points for this cube */
  get points() {
    return this._cuboid.points
  }

  /**
   * Checks if the vector is within this cube
   *
   * @param {Vector3} vector3
   * @return {boolean}
   */
  contains(vector3) {
    return this._cuboid.contains(vector3)
  }

  /**
   * Checks if the cubes are equal
   *
   * @param {CubeRegion} other
   * @return {boolean}
   */
  equals(other) {
    return other && _.isEqual(this._cuboid, other._cuboid)
  }
}
